### PR TITLE
USB pumps

### PR DESCRIPTION
**Pull request by Ghilker**
## About The Pull Request
Add USB port to the atmos pumps (the pipe one)

## Why It's Good For The Game
Expansion of existing content

## Changelog

:cl:
expansion: Add USB port to the atmos pumps
/:cl:

### CODE REVIEW

File: code/modules/atmospherics/machinery/components/binary_devices/pump.dm [hunk 1]
```diff
@@ -180,3 +189,78 @@
...
+	var/datum/port/input/request_data
+
+	var/datum/port/output/input_pressure
+	var/datum/port/output/output_pressure
+	var/datum/port/output/input_temperature
+	var/datum/port/output/output_temperature
+
+	var/datum/port/output/is_active
+	var/datum/port/output/turned_on
+	var/datum/port/output/turned_off
```
Comment (Watermelon914) [inline code/modules/atmospherics/machinery/components/binary_devices/pump.dm:209]: document these variables

File: code/modules/atmospherics/machinery/components/binary_devices/pump.dm [hunk 2]
```diff
@@ -180,3 +189,78 @@
...
+
+	var/obj/machinery/atmospherics/components/binary/pump/connected_pump
+
+/obj/item/circuit_component/atmos_pump/populate_ports()
+	pressure_value = add_input_port("New Pressure", PORT_TYPE_NUMBER)
```
[the remark this answers is not in the document]
Comment (Watermelon914) [reply code/modules/atmospherics/machinery/components/binary_devices/pump.dm:226]: If you do this, in your new proc, add CIRCUIT_TRIGGER to the very start of the proc.

File: code/modules/atmospherics/machinery/components/binary_devices/pump.dm [hunk 3]
```diff
@@ -180,3 +189,93 @@
...
+	if(off ==  port)
+		connected_pump.set_on(FALSE)
+		return
+	if(request_data == port)
+		var/datum/gas_mixture/air_input = connected_pump.airs[1]
+		var/datum/gas_mixture/air_output = connected_pump.airs[2]
+		input_pressure.set_output(air_input.return_pressure())
+		output_pressure.set_output(air_output.return_pressure())
+		input_temperature.set_output(air_input.return_temperature())
+		output_temperature.set_output(air_output.return_temperature())
```
Comment (Watermelon914) [inline code/modules/atmospherics/machinery/components/binary_devices/pump.dm:281]: you can move these all to their own triggers as well. It's ideal if everything isn't in input_received

Comment (Ghilker) [reply code/modules/atmospherics/machinery/components/binary_devices/pump.dm:281]: aight!

**Comment by Pickle-Coding** [comment]
Can't wait for the day when we can create fully automated HFRs.

File: code/modules/atmospherics/machinery/components/binary_devices/pump.dm [hunk 4]
```diff
@@ -180,3 +189,97 @@
...
+		connected_pump = shell
+		RegisterSignal(connected_pump, COMSIG_PUMP_SET_ON, .proc/handle_pump_activation)
+
+/obj/item/circuit_component/atmos_pump/unregister_usb_parent(atom/movable/shell)
+	UnregisterSignal(connected_pump, COMSIG_PUMP_SET_ON)
+	connected_pump = null
+	return ..()
+
+/obj/item/circuit_component/atmos_pump/proc/handle_pump_activation(datum/source, active)
+	is_active.set_output(active)
```
Comment (Watermelon914) [inline code/modules/atmospherics/machinery/components/binary_devices/pump.dm:252]: SIGNAL_HANDLER here

Comment (Ghilker) [reply code/modules/atmospherics/machinery/components/binary_devices/pump.dm:252]:
```suggestion
	SIGNAL_HANDLER
	is_active.set_output(active)
```

**Pull request merged**